File: generated_code/cpp/solution_3556_1.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
private:
    const int MOD = 1e9 + 7;

    // Helper function for modular exponentiation: (base^exp) % MOD
    long long power(long long base, long long exp) {
        long long res = 1;
        base %= MOD;
        while (exp > 0) {
            if (exp % 2 == 1) res = (res * base) % MOD;
            base = (base * base) % MOD;
            exp /= 2;
        }
        return res;
    }

public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        // Edge case: if multiplier is 1, array elements never change value
        if (multiplier == 1) {
            vector<int> res(nums.size());
            for (size_t i = 0; i < nums.size(); ++i) {
                res[i] = nums[i] % MOD;
            }
            return res;
        }

        int n = nums.size();
        // Min-heap to track {value, original_index}. 
        // Pairs are compared by value first, then by index to handle duplicates correctly.
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
        
        long long max_val = 0;
        for (int i = 0; i < n; ++i) {
            pq.push({nums[i], i});
            max_val = max(max_val, (long long)nums[i]);
        }

        // Phase 1: Simulate operations greedily until the minimum element 
        // multiplied by the multiplier exceeds the initial maximum element.
        // This brings all elements to a state where they are roughly balanced.
        while (k > 0 && pq.top().first * multiplier <= max_val) {
            auto [val, idx] = pq.top();
            pq.pop();
            pq.push({val * multiplier, idx});
            k--;
        }

        // Phase 2: Once the array is balanced, additional cycles of operations 
        // will multiply each element by `multiplier` in a predictable, uniform way.
        int full_cycles = k / n;
        int remaining_ops = k % n;

        // Precompute multiplier factors using modular exponentiation
        long long factor_full = power(multiplier, full_cycles);
        long long factor_extra = power(multiplier, full_cycles + 1);

        vector<int> result(n);
        
        // The first `remaining_ops` elements extracted from the heap will get 
        // an extra multiplication since they are the smallest elements in the current state.
        for (int i = 0; i < n; ++i) {
            auto [val, idx] = pq.top();
            pq.pop();
            
            long long current_factor = (i < remaining_ops) ? factor_extra : factor_full;
            result[idx] = ((val % MOD) * current_factor) % MOD;
        }

        return result;
    }
};

File: generated_code/python/solution_3556_1.py
from typing import List
import heapq

class Solution:
    def getFinalState(self, nums: List[int], k: int, multiplier: int) -> List[int]:
        # Edge case: if multiplier is 1, operations don't change the values
        if multiplier == 1:
            return [x % 1000000007 for x in nums]
            
        MOD = 10**9 + 7
        n = len(nums)
        
        # Min-heap stores tuples of (value, original_index)
        # Python's heapq handles tuples by comparing elements from left to right,
        # which perfectly matches the requirement: sort by value, then by earliest index.
        heap = [(val, i) for i, val in enumerate(nums)]
        heapq.heapify(heap)
        
        # Step 1: Simulate operations until the maximum element in nums 
        # becomes less than or equal to the minimum element * multiplier.
        # This state ensures that all elements have entered a relative cycle where
        # operations will cycle through all elements uniformly.
        max_val = max(nums)
        
        while k > 0 and heap[0][0] * multiplier <= max_val:
            val, idx = heapq.heappop(heap)
            new_val = val * multiplier
            nums[idx] = new_val
            heapq.heappush(heap, (new_val, idx))
            k -= 1
            
        # Step 2: If there are still operations left, they can be distributed 
        # evenly across all elements using fast exponentiation.
        if k > 0:
            # Each element gets at least q additional multiplier applications
            q = k // n
            # The first r elements in the heap get one extra multiplier application
            r = k % n
            
            # Fast exponentiation for the multiplier powers
            pow_q = pow(multiplier, q, MOD)
            pow_q_plus_1 = (pow_q * multiplier) % MOD
            
            # Extract elements from the heap in their exact sorted order of simulation
            sorted_elements = []
            while heap:
                sorted_elements.append(heapq.heappop(heap))
                
            # Apply the calculated powers to the elements based on their cycle position
            for i, (val, idx) in enumerate(sorted_elements):
                if i < r:
                    nums[idx] = (val % MOD * pow_q_plus_1) % MOD
                else:
                    nums[idx] = (val % MOD * pow_q) % MOD
        else:
            # If k reached 0 during Step 1, just apply modulo to all elements
            for i in range(n):
                nums[i] %= MOD
                
        return nums